rust/xaynet-client/src/state_machine/phases/phase.rs

```diff
@@ -0,0 +1,212 @@
+use derive_more::From;
+use thiserror::Error;
+
+use xaynet_core::{
+    common::RoundParameters,
+    crypto::SigningKeyPair,
+    mask::MaskConfig,
+    message::Payload,
+};
+
+use super::{Awaiting, NewRound, Sum, Sum2, Update};
+use crate::{
+    settings::{MaxMessageSize, Settings},
+    state_machine::{io::StateMachineIO, StateMachine, TransitionOutcome},
+    MessageEncoder,
+};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct State<P> {
+    /// data specific to the current phase
+    pub phase: P,
+    /// data common to most of the phases
+    pub shared: SharedState,
+}
+
+impl<P> State<P> {
+    pub fn new(shared: SharedState, phase: P) -> Self {
+        Self { shared, phase }
+    }
+}
+
+#[derive(Debug)]
+pub struct Phase<P, IO> {
+    pub state: State<P>,
+    /// Opaque client for performing IO tasks: talking with the
+    /// coordinator API, loading models, etc.
+    pub io: IO,
+}
+
+/// Store for all the data that are common to all the phases
+#[derive(Serialize, Deserialize, Debug)]
+pub struct SharedState {
+    pub keys: SigningKeyPair,
+    pub mask_config: MaskConfig,
+    pub scalar: f64,
+    pub message_size: MaxMessageSize,
+    pub round_params: RoundParameters,
+}
+
+impl SharedState {
+    pub fn new(settings: Settings) -> Self {
+        Self {
+            keys: settings.keys,
+            mask_config: settings.mask_config,
+            scalar: settings.scalar,
+            message_size: settings.max_message_size,
+            round_params: RoundParameters::default(),
+        }
+    }
+}
+
+/// Represent an atomic step performed by a phase. If the step results
+/// in progress being made, the resulting updated state machine is
+/// returned as `Progress::Some`.
```

Comment:
Not sure but I think it should be 
```suggestion
/// returned as `TransitionOutcome::Complete`.
```